src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -261,12 +261,16 @@ Symbol* SystemDictionary::class_name_symbol(const char* name, Symbol* exception,
   if (name == nullptr) {
     THROW_MSG_NULL(exception, "No class name given");
   }
-  if ((int)strlen(name) > Symbol::max_length()) {
+  if (strlen(name) > static_cast<size_t>(Symbol::max_length())) {
     // It's impossible to create this class;  the name cannot fit
     // into the constant pool.
+    // To avoid internal snprintf INT_MAX limit we reduce the maximum length
+    // of name to print, by the length of the rest of the formatted message.
+    int print_limit = INT_MAX - 45;
```

Comment:
It does. It is the text up to the %s and the expanded %d which will be 65535, plus one. The plus one deals with a seeming off-by-one bug with macOS vsnprintf.

Reply:
But I'm realizing now that this should not work ... it needs to account for the actual strlen of the name.

Reply:
No I take that back.

Reply:
This seems fine.  You say what the 45 is for.

Reply:
Do you really want to throw an exception with a message length of up to 2 GB? This may find its way into logfiles etc. Even printing out 64K worth of name may be a bit much.

Reply:
But `Exceptions::fthrow` truncates at 1024 bytes anyway, which I think is absolutely fine. So the print_limit could just be 1024.

Reply:
It could ... but that limit of `fthrow` is not really part of the API or exposed. It is also a somewhat arbitrary VM limit and the same kind of check in Java code would just include the complete name regardless of size (limited to `int_max` latin1 characters of course). Even at 64K you may want to see why your class name is unexpectedly longer than you thought. And for other callers of this API the truncation applied by `fthrow` might be considered a bug as it would impede debugging.

I wanted to avoid the `os::vsnprintf` problem without imposing any other policy choice here.

Reply:
PS. Thanks for taking a look!

Reply:
I get what you mean, but then, we should take the limit you impose here (print_limit) at face value and act as if Exceptions::fthrow would not impose any limits (since we cannot know them since they are implementation details).

So, again there is the question if we want to pay 2 GB (multiple times, since we need a buffer for vsprintf to assemble it, then memory in java heap for the resulting j.l.String). Something akin to your log messages where we omit the middle of the string may serve us here as well. Or, just a bigger but still reasonable limit (e.g. 512 KB).

Reply:
You're right. I can apply the "abridged" pattern here as done in `java_lang_String::print` using the `MaxStringPrintSize` global. But this is now encroaching on the remedial actions I planned to undertake for [JDK-8328882](https://bugs.openjdk.org/browse/JDK-8328882).  I only need this change in this PR to handle the test case for the actual bug fix (otherwise I have to drop the part of the test that verifies it is the expected error message).

I will look at creating a sub-task of JDK-8328882 to handle callers of `Exceptions::fthrow` and fix this issue under that task first.

Reply:
Well I created a sub-task as I indicated and it turned out that most places are safe but can still have "names" up to 64K, but the nature of the error messages are such that they are not really amenable to abridged versions, plus you may need the full name to actually see the error. So in that regard I think `fthrow` is actually wrong to impose its own 1K limit as it will prevent you from debugging things - in some cases long before you get anywhere near excessive individual names.

But back to the individual case at hand ... given we know its a giant name and its going to get truncated anyway, then the abridged version is better IMO as I still contend that given it is an erroneous condition it is most likely the error occurred at the start or end (I've absolutely no data to back that up : ) ).

Reply:
I think most people will be helped even with the abridged error messages. Better than having no error text at all.